core/lib/dal/src/blocks_dal.rs

```diff
@@ -549,14 +556,131 @@ impl BlocksDal<'_, '_> {
         Ok(storage_oracle_info.and_then(DbStorageOracleInfo::into_optional_batch_oracle_info))
     }
 
+    /// The server is doing precommits based on the txs.
+    /// That means, that the last fictive l2 block will never be included into precommit txs and it's the only way how the miniblock will have no txs.
+    /// So we have to check if the last 2 rolling txs hashes for miniblocks are equal and if yes, we can set the final precommit tx id for the batch,
+    /// and that will mean we can commit the batch.
+    pub async fn set_final_precommit_operation(&mut self) -> DalResult<()> {
+        let rows = sqlx::query!(
+            r#"
+            SELECT
+                l1_batches.number AS l1_batch_number,
+                miniblocks.eth_precommit_tx_id,
+                miniblocks.rolling_txs_hash,
+                miniblocks.number AS miniblock_number
+            FROM
+                l1_batches
+            JOIN LATERAL (
+                SELECT
+                    eth_precommit_tx_id,
+                    rolling_txs_hash,
+                    number
+                FROM miniblocks
+                WHERE l1_batch_number = l1_batches.number
+                ORDER BY number DESC
+                LIMIT 2
+            ) miniblocks ON TRUE
+            WHERE
+                l1_batches.number > 0
+                AND l1_batches.is_sealed
+                AND l1_batches.eth_commit_tx_id IS NULL
+                AND l1_batches.final_precommit_eth_tx_id IS NULL
```

Comment:
```suggestion
                AND l1_batches.final_precommit_eth_tx_id IS NULL
                AND miniblocks.rolling_txs_hash IS NOT NULL
```
I think we can filter batches without precommits enabled this way. 

Overall I don't like it, becouse you perform a very complex SELECT that "who knows how long takes on production" that potentially returns tons of data. To me its a cointoss if this will be actually fine. We have some query statistics right? I would make sure this doesn't spike when this feature is enabled on a big chain.

Reply:
This query is not dangerous. 
It retuns pretty low amount of data, normally we don't have more than 100 uncommitted batches. 
this query is called only when the feature is enabled. meanwhile. i'm not against adding this check